Convert net usage to annual nets distributed per capita

norasIDM opened the pull request:
Hi Pete, can you let me know if this is sort of what you had in mind for aim 1? I am also unsure to what degree we want to rely on reading in data from other sources or saving them here.

norasIDM committed:
Working on extrapolation

norasIDM committed:
Exploring extrapolation options

norasIDM committed:
Assume linear extrapolation of Loess curve

norasIDM committed:
Updated 2020 dataset for Loess curve

norasIDM committed:
error fixed

norasIDM committed:
Conversion from NPC to nets distributed per year

norasIDM committed:
Function to convert usage into annual nets distributed per capita

norasIDM committed:
Adapted for package

norasIDM committed:
Package functionality and reverse function of distribution=>usage conversion added

pwinskill commented on R/prepare_data.R:
```diff
@@ -0,0 +1,50 @@
+#' @importFrom utils 'read.csv'
+#' @export
+
+prepare_data <- function() {
```
Can we put all variables that the user might want to change, or that might have updated defaults in the future as arguments for this function.
e.g. data location address strings, year filters etc.

norasIDM replied:
I broke up the functions into the individual conversion steps and added some tests for each. Documentation should be up-to-date

pwinskill commented on R/usage_to_annual_nets_distributed.R:
```diff
@@ -0,0 +1,234 @@
+  
+  # Read in datasets of country-specific use rate and the Loess curve of 
+  # access vs nets per capita
+  data <- prepare_data()
```
I think it might be best to have this `prepare_data()` call outside of this function then passed through as an argument. Otherwise a user might make lots of individual calls to `find_annual_nets_distributed()` and each time it is downloading and preparingt the whole MAP dataset.

pwinskill commented on R/usage_to_annual_nets_distributed.R:
```diff
@@ -0,0 +1,234 @@
+#' @export
+
+find_annual_nets_distributed <- function(target_usage, country_iso3="all", 
+                                         extrapolate_npc = "linear", k=20) {
```
This seems a bit pedantic (sorry!) but good to follow a consistent style guide (e.g. spaces arounf <- and =).
The tidyverse style guide is a good reference here: https://style.tidyverse.org/index.html

pwinskill commented on R/usage_to_annual_nets_distributed.R:
```diff
@@ -0,0 +1,234 @@
+    stop("Target usage(s) cannot be achieved with estimated use rate in any of the countries. Refer to maximum usage above.")
+  } else if(any(target_usage>min(country_use_rate[country_use_rate$iso3 %in% country_iso3,"use_rate"]))) {
+    print(max_usage)
+    print("Target usage(s) cannot be achieved with estimated use rate in some of the countries - return NA. Refer to maximum usage above.")
```
Replace these print statements with warnings `warning()` as that will play nicely with formal error checking/handling

norasIDM replied:
Replaced with warning() where appropriate (some are more information rather than warnings)

pwinskill commented on R/usage_to_annual_nets_distributed.R:
```diff
@@ -0,0 +1,234 @@
+#' @importFrom stats 'approx' 'reshape' 'loess' 'predict'
+#' @export
+
+find_annual_nets_distributed <- function(target_usage, country_iso3="all", 
```
for consistency can `country_iso3` be renamed `iso3`

norasIDM replied:
Data now read in externally so country selection will also be done at that stage

pwinskill commented on R/usage_to_annual_nets_distributed.R:
```diff
@@ -0,0 +1,234 @@
+    country_iso3 <- country_use_rate$iso3
+  }
+  
+  if(extrapolate_npc == "linear") {
```
"linear" here makes it seem like there will be other extrapolation options. I'd just have the argument `extrapolate_npc` as boolean, with an explainer in the documetnation that if set to TRUE the extrapolation will be linear

norasIDM replied:
Added an additional option

pwinskill commented on R/usage_to_annual_nets_distributed.R:
```diff
@@ -0,0 +1,234 @@
+    loess_for_prediction <- rbind(loess_for_prediction,
+                                  data.frame(iso3=NA, month=NA, access=1, 
+                                             percapita_nets = 1, loess_predicted_access=1))
+    print("Access-nets per capita curve is extrapolated linearly beyond observed levels")
```
As below comment (`warning()`)

norasIDM replied:
I did not change this to warning() since it is more of a reminder of what is inputted

pwinskill commented on R/usage_to_annual_nets_distributed.R:
```diff
@@ -0,0 +1,234 @@
+                                             percapita_nets = 1, loess_predicted_access=1))
+    print("Access-nets per capita curve is extrapolated linearly beyond observed levels")
+  } else {
+    print("No extrapolation beyond observed access-nets per capita relationship - return NA")
```
Not necessary to have this

pwinskill commented on R/usage_to_annual_nets_distributed.R:
```diff
@@ -0,0 +1,234 @@
+#' @param target_usage A single value or vector of desired target usages to model.
+#' @param country_iso3 Vector of ISO3 codes of African countries for which to return results. 
+#' Default = "all".
+#' @param extrapolate_npc Option to extrapolate target nets per capita beyond fitted 
```
For the extrapolation, I think we want the trend from the end of curve (say the last 4 points) to be extrapolated.

Currently if I run:
```
p1 <- find_annual_nets_distributed(seq(0, 1, 0.01), country_iso3 = "NGA")
plot(p1$annual_percapita_nets_distributed ~ p1$target_use,
     xlab = "Target Use", ylab = "Annual nets distributed")
```
I get:
![image](https://user-images.githubusercontent.com/16702715/151959307-81250258-ad9e-4b40-8e70-314bc014ddb5.png)

This makes it get more efficient at the end, which I don't think would happen

norasIDM replied:
I included this unrealistic option in case you wanted to reach an access of 1 (so you can run higher usages) but let me know if you think its better to remove altogether. Option "loess" now added is more realistic

pwinskill commented on R/usage_to_annual_nets_distributed.R:
```diff
@@ -0,0 +1,234 @@
+  half_life_data <- data$half_life_data
+  
+  # Estimate time at which all nets are lost (parameter l) based on net half life in each country
+  half_life_data$l <- half_life_data$half_life/sqrt(1 - k / (k - log(0.5)))
```
This is quite a large function and a number of sections can be split out into their own small functions that perform single, simple tasks. These are then much easier to unit test.

pwinskill commented on R/usage_to_annual_nets_distributed.R:
```diff
@@ -0,0 +1,234 @@
+                             "target_percapita_nets","target_access", 
+                             "target_usage")])
+  
+  # TO DO:
```
Good to keep comments like this out of the code - they have a tendancy to sneek through to propoer releases. Feel free to stick stuff like this in the issues + conmversations on those issues.

Extrpolating at the lower end would be helpful yes.
I think we can have percapita nets >1 (its a bit crazy, but not impossible)

pwinskill commented on R/usage_to_annual_nets_distributed.R:
```diff
@@ -0,0 +1,234 @@
+#' @param k Fixed rate. Default = 20 from Bertozzi-Villa, Amelia, et al. Nature communications 12.1 (2021): 1-12.
+#' @param l Country-specific time at which no nets are retained (nets=0).
+#' @export
+integrate_net_loss <- function(k,l) {
```
We probably want at least the upper to be an argument here  - we might want to assume other distribution intervals than 3 yearly.

pwinskill commented on R/usage_to_annual_nets_distributed.R:
```diff
@@ -0,0 +1,234 @@
+  # is distributed nets every 3 years (this is equivalent to a smoothed 3 yearly whole 
+  # population distribution).
+  nets_distributed$annual_percapita_nets_distributed <- nets_distributed$target_percapita_nets * 
+    integrate_net_loss_vector(k=k, l = nets_distributed$l) * (1/3)
```
This conversion is assuming equilibrium distributions of equial size on a regular schedule.

Eventually I think we might want to "fit" the annual_percapita_nets_distributed to match the target_per_capita nets over time. This is an (almost) identical operation to the fitting the malariasimulation input distirbutions to match target population usage over time - so we should be able to use the same functionality.

pwinskill replied:
I will add this as an issue

pwinskill reviewed (changes requested):
This is going to be so helpful!

Lots of suggestions in specific comments.

Some general suggestions:
1. Lots of the functions remain large. It will be best to break them down as much as possible into small chunks where each function does a single small "job"
2. Doing (1) will also us to include unit tests of these small functions much more easily. We will need to add quite a few tests here to make sure the code is doing what we want - especially as we have quite a few edge cases.

I had some issues with the documentation. Worth making sure you've run `devtools::document()` to update before pushing any commits.

Commit:
Restructuring

norasIDM committed:
prepare_data inputs

norasIDM committed:
Vignette

norasIDM committed:
Vignette updates

norasIDM committed:
Tests added

norasIDM committed:
Revert to previous

inescgu committed:
Conversion of annual nets distributed to per capita bed net usage.

Reverse of the conversion method described in the usage to npc function. Experiences some functionality issues with unstable internet connections - these can be solved by calling prepare_data() beforehand.

inescgu committed:
Merge branch 'usage_to_npc' of https://github.com/mrc-ide/netz into usage_to_npc

inescgu committed:
Adding documentation to nets_to_usage functionality.

norasIDM committed:
Tests added

norasIDM committed:
Documentation change

norasIDM committed:
Test updated

inescgu committed:
Renamed npc_to_usage to nets_to_usage, added test file.

inescgu committed:
Delete npc_to_usage.R

inescgu committed:
Merge branch 'usage_to_npc' of https://github.com/mrc-ide/netz into usage_to_npc

inescgu committed:
Updates on input checks.
